cmd/wasm-client/main.go

```diff
@@ -281,6 +313,88 @@ func (w *wasmClient) InvokeRPC(_ js.Value, args []js.Value) interface{} {
 
 }
 
+func (w *wasmClient) GetExpiry(_ js.Value, _ []js.Value) interface{} {
+	if w.mac == nil {
+		return nil
+	}
+
+	expiry, found := checkers.ExpiryTime(nil, w.mac.Caveats())
+	if !found {
+		return nil
+	}
+
+	return js.ValueOf(expiry.Unix())
+}
+
+func (w *wasmClient) HasPermissions(_ js.Value, args []js.Value) interface{} {
+	if len(args) != 1 {
+		return js.ValueOf(false)
+	}
+
+	if w.mac == nil {
+		return js.ValueOf(false)
+	}
+
+	// Convert JSON callback to grpc URI. JSON callbacks are of the form:
+	// `lnrpc.Lightning.WalletBalance` and the corresponding grpc URI is of
+	// the form: `/lnrpc.Lightning/WalletBalance`. So to convert the one to
+	// the other, we first convert all the `.` into `/`. Then we replace the
+	// first `/` back to a `.` and then we prepend the result with a `/`.
+	rg := regexp.MustCompile("(\\w+)\\.(\\w+)\\.(\\w+)")
+	uri := rg.ReplaceAllString(args[0].String(), "/$1.$2/$3")
+
+	ops, ok := perms[uri]
+	if !ok {
+		log.Errorf("uri %s not found in known permissions list", uri)
+		return js.ValueOf(false)
+	}
+
+	rawID := w.mac.Id()
+	if rawID[0] != byte(bakery.LatestVersion) {
+		log.Errorf("invalid macaroon version: %x", rawID)
+		return js.ValueOf(false)
+	}
+
+	decodedID := &lnrpc.MacaroonId{}
+	idProto := rawID[1:]
+	err := proto.Unmarshal(idProto, decodedID)
+	if err != nil {
+		log.Errorf("unable to decode macaroon: %v", err)
+		return js.ValueOf(false)
+	}
+
+	// Ensure that the macaroon contains each of the required permissions
+	// for the given URI.
+	var found bool
+	for _, op := range ops {
+		found = false
+		for _, macOp := range decodedID.Ops {
```

Comment:
Hmm, maybe extract this part into a function so things become easier to read?

Reply:
updated 👍  also added a lookup map that makes things way simpler